python/ray/experimental/state/state_manager.py

```diff
@@ -0,0 +1,191 @@
+import inspect
+
+from functools import wraps
+
+import grpc
+import ray
+
+from typing import Dict, List
+
+from ray.core.generated.gcs_service_pb2 import (
+    GetAllActorInfoRequest,
+    GetAllActorInfoReply,
+    GetAllPlacementGroupRequest,
+    GetAllPlacementGroupReply,
+    GetAllNodeInfoRequest,
+    GetAllNodeInfoReply,
+    GetAllWorkerInfoRequest,
+    GetAllWorkerInfoReply,
+)
+from ray.core.generated.node_manager_pb2 import (
+    GetTasksInfoRequest,
+    GetTasksInfoReply,
+    GetNodeStatsRequest,
+    GetNodeStatsReply,
+)
+from ray.core.generated import gcs_service_pb2_grpc
+from ray.core.generated.node_manager_pb2_grpc import NodeManagerServiceStub
+from ray.core.generated.reporter_pb2_grpc import ReporterServiceStub
+from ray.dashboard.modules.job.common import JobInfoStorageClient, JobInfo
+
+
+class StateSourceNetworkException(Exception):
+    """Exceptions raised when there's a network error from data source query."""
+
+    pass
+
+
+def apply_state_client_api_policy(func):
+    """Apply the basic API policy to each APIs, such as retry or exception policies.
+
+    It is a helper method for `StateDataSourceClient`.
+    """
+    assert inspect.iscoroutinefunction(func)
+
+    @wraps(func)
+    async def func_with_api_policy(*args, **kwargs):
+        # TODO(sang): Add a retry policy.
+        try:
+            return await func(*args, **kwargs)
+        except (
+            # https://grpc.github.io/grpc/python/grpc_asyncio.html#grpc-exceptions
+            grpc.aio.AioRpcError,
+            grpc.aio.InternalError,
+            grpc.aio.AbortError,
+            grpc.aio.BaseError,
+            grpc.aio.UsageError,
+        ) as e:
+            raise StateSourceNetworkException(
```

Comment:
Also debating if we need this. This will help us control what exceptions are thrown from the APis, which will be helpful for unit testing